src/glob.rs

```diff
@@ -677,6 +750,140 @@ mod test {
     );
   }
 
+  #[test]
+  fn file_patterns_split_by_base_dir_unexcluded() {
+    let temp_dir = TempDir::new().unwrap();
+    let patterns = FilePatterns {
+      base: temp_dir.path().to_path_buf(),
+      include: None,
+      exclude: PathOrPatternSet::new(vec![
+        PathOrPattern::from_relative(temp_dir.path(), "./ignored").unwrap(),
+        PathOrPattern::from_relative(temp_dir.path(), "!./ignored/unexcluded")
+          .unwrap(),
+        PathOrPattern::from_relative(temp_dir.path(), "!./ignored/test/**")
+          .unwrap(),
+      ]),
+    };
+    let split = ComparableFilePatterns::from_split(
+      temp_dir.path(),
+      &patterns.split_by_base(),
+    );
+    assert_eq!(
+      split,
+      vec![
+        ComparableFilePatterns {
+          base: "ignored/unexcluded".to_string(),
+          include: None,
+          exclude: vec![
+            // keep the glob for the current dir because it
+            // could be used to override the .gitignore
+            "!ignored/unexcluded".to_string(),
+          ],
+        },
+        ComparableFilePatterns {
+          base: "ignored/test".to_string(),
+          include: None,
+          exclude: vec!["!ignored/test/**".to_string(),],
+        },
+        ComparableFilePatterns {
+          base: "".to_string(),
+          include: None,
+          exclude: vec![
+            "ignored".to_string(),
+            "!ignored/unexcluded".to_string(),
+            "!ignored/test/**".to_string(),
+          ],
+        },
```

Comment:
The vector here is what the CLI uses to walk the directories. So first it goes into ignored/unexcluded, then ignored/test, then the root directory. The CLI's walking prevents entering a directory after it's already been walked.